alvr/server/cpp/platform/linux/EncodePipelineVAAPI.cpp

```diff
@@ -196,6 +196,8 @@ alvr::EncodePipelineVAAPI::EncodePipelineVAAPI(Renderer *render, VkContext &vk_c
   encoder_ctx->rc_max_rate = encoder_ctx->bit_rate;
   encoder_ctx->rc_buffer_size = encoder_ctx->bit_rate / settings.m_refreshRate;
 
+  encoder_ctx->flags |= AV_CODEC_FLAG_GLOBAL_HEADER;
```

Comment:
This doesn't work with mesa drivers.

Reply:
Huh, I see someone patched global headers for h264 this in private repo, but it's still not in mesa?
https://github.com/sgothel/mesa/commit/5e10db6f3f22a7d687ad7eeded3bbf0866978628

Reply:
Yes, it's not. I think it's time to force static ffmpeg with custom patches on linux builds :)

Reply:
how about we config guard it?

Reply:
what patch to ffmpeg is needed?

Reply:
Something like this (untested), but it still won't work correctly.

With this flag VAAPI shouldn't output headers in the bitstream, but instead FFmpeg should write them in the extradata buffer. But since mesa/radeonsi doesn't support this flag the headers will also be present in the output bitstream (in every i/idr frame) and possibly slightly different from the headers in extradata.

```patch
From 03e6b5617c94bf6073a17a0a7054e70345b512ae Mon Sep 17 00:00:00 2001
From: David Rosca <nowrep@gmail.com>
Date: Sat, 1 Jul 2023 10:35:16 +0200
Subject: [PATCH] vaapi_encode: Force enable global headers

---
 libavcodec/vaapi_encode.c | 3 +--
 1 file changed, 1 insertion(+), 2 deletions(-)

diff --git a/libavcodec/vaapi_encode.c b/libavcodec/vaapi_encode.c
index bfca315a7a..ecce9bd721 100644
--- a/libavcodec/vaapi_encode.c
+++ b/libavcodec/vaapi_encode.c
@@ -2719,8 +2719,7 @@ av_cold int ff_vaapi_encode_init(AVCodecContext *avctx)
         }
     }
 
-    if (ctx->va_packed_headers & VA_ENC_PACKED_HEADER_SEQUENCE &&
-        ctx->codec->write_sequence_header &&
+    if (ctx->codec->write_sequence_header &&
         avctx->flags & AV_CODEC_FLAG_GLOBAL_HEADER) {
         char data[MAX_PARAM_BUFFER_SIZE];
         size_t bit_len = 8 * sizeof(data);
-- 
2.41.0


```

Reply:
But we need to patch the driver, not ffmpeg, no? I've mentioned to the commit that adds VA_ENC_PACKED_HEADER_SEQUENCE to mesa, so ffmpeg should be working out of the box, isn't it?

Reply:
Well ALVR can only ship patched ffmpeg, not mesa driver.

Reply:
mesa patch:

```patch
From 6bd74eae97dc73da5892996031426845cef8558d Mon Sep 17 00:00:00 2001
From: David Rosca <nowrep@gmail.com>
Date: Sat, 1 Jul 2023 11:24:37 +0200
Subject: [PATCH] frontends/va: Advertise PACKED_HEADER_SEQUENCE support for
 AVC

---
 src/gallium/frontends/va/config.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/src/gallium/frontends/va/config.c b/src/gallium/frontends/va/config.c
index 0e484763c..84742148e 100644
--- a/src/gallium/frontends/va/config.c
+++ b/src/gallium/frontends/va/config.c
@@ -235,6 +235,8 @@ vlVaGetConfigAttributes(VADriverContextP ctx, VAProfile profile, VAEntrypoint en
             break;
          case VAConfigAttribEncPackedHeaders:
             value = VA_ENC_PACKED_HEADER_NONE;
+            if ((u_reduce_video_profile(ProfileToPipe(profile)) == PIPE_VIDEO_FORMAT_MPEG4_AVC))
+               value |= VA_ENC_PACKED_HEADER_SEQUENCE;
             if ((u_reduce_video_profile(ProfileToPipe(profile)) == PIPE_VIDEO_FORMAT_HEVC))
                value |= VA_ENC_PACKED_HEADER_SEQUENCE;
             else if (u_reduce_video_profile(ProfileToPipe(profile)) == PIPE_VIDEO_FORMAT_AV1)
@@ -620,6 +622,8 @@ vlVaCreateConfig(VADriverContextP ctx, VAProfile profile, VAEntrypoint entrypoin
       if (attrib_list[i].type == VAConfigAttribEncPackedHeaders) {
          if (config->entrypoint != PIPE_VIDEO_ENTRYPOINT_ENCODE ||
              (((attrib_list[i].value != 0)) &&
+              ((attrib_list[i].value != 1) || u_reduce_video_profile(ProfileToPipe(profile))
+               != PIPE_VIDEO_FORMAT_MPEG4_AVC) &&
               ((attrib_list[i].value != 1) || u_reduce_video_profile(ProfileToPipe(profile))
                != PIPE_VIDEO_FORMAT_HEVC) &&
               ((attrib_list[i].value != 3) || u_reduce_video_profile(ProfileToPipe(profile))
-- 
2.41.0

```

But I still think it should be patched in ffmpeg because it doesn't really need the packed headers support to output global headers. Gstreamer also works fine without it.

And advertising the support in mesa is a lie anyway, because the driver will ignore it and output the headers by itself (which is also true for hevc/av1, although there it at least actually parses the headers and use some parameters).

Reply:
It's now supported in mesa git.

Reply:
Awesome anything blocking it now other than the ci fail?

Reply:
Should be good to go after #1751